### PR TITLE
chore: enhance custom error handling and documentation

### PR DESCRIPTION
**Pull request by FlynnOConnell**
- Refactored `CustomError` to handle specific PostgreSQL error types, providing clearer error messages for database violations.
- Updated the `IntoResponse` implementation for `CustomError` to ensure appropriate HTTP status codes and error messages are returned in API responses.
- Added documentation for both the `From<SqlxError>` and `IntoResponse` trait implementations.

## Errors to Handle

1. Users

- `GET user` -> `Path(user_id)` | `40X`
    - if client sends an incorrect or unauthorized request
- ` CREATE user` ->  `Json(credentials)`
    - duplicate username -> `sqlx::Error` | `409` + `Location` header -> existing/conflicting resource? 
    - password not strong enough -> not `sqlx::Error` | `403 Forbidden` ? 
- `DELETE user` -> `Path(user_id)` | `40X`
    - `user_id` doesn't exist shouldn't ever happen
- `UPDATE user` -> `Path(user_id)`
    - Right now, empty first/last name and password work, may want to change this -> `40X`
    - `user_id` doesn't exist shouldn't ever happen | `40X`
- `Signup` and `Login` user should be a bigger separate issue

2. Channels

- All follow mostly the same logic as users for status codes and error handling

3. Messages

- In theory, `Path(user_id)`, `Path(channel_id)` and `Path(message_id)` references shouldn't be an issue on our end | `40X`

## Tasks: 

- [x] Better handle message references after `user` or `channel` deletion, Addresses #52

Still more work to be done on error handling. There are more cases to cover and potentially a better implementation. Consider a more robust approach [as is done here](https://github.com/davidpdrsn/realworld-axum-sqlx/blob/main/src/http/error.rs) for the future.

### CODE REVIEW

**Comment by FlynnOConnell** [comment]
Creating a user that already exists gives: 

```json
{
    "error": "Unique violation: duplicate key value violates unique constraint \"fluke_user_email_key\""
}
```

Creating a message with an invalid `user_id` gives:

```json
{
    "error": "Foreign violation: insert or update on table \"message\" violates foreign key constraint \"fk_message_user_id\""
}
```
Though this shouldn't really happen. Many of the errors shouldn't happen but are logged anyway.

File: backend/src/errors.rs [hunk 1]
```diff
@@ -3,89 +3,94 @@ use axum::{
...
-            SignupError::NonUniqueIdError => {
-                write!(f, "Duplicate email contained a duplicate key.")
-            }
-            SignupError::UnknownQueryError => {
-                write!(f, "Database query contained an unspecified error.")
-            }
-            SignupError::UnknownDatabaseError => write!(f, "Database error, not query related."),
-        }
-    }
-}
```
Comment (FlynnOConnell) [inline backend/src/errors.rs:91]: This code felt redundant. and for now is unused. We can merge into the CustomError, and we can easily add more custom error names for the struct like SignupError, EmailNotFound, etc.

File: backend/src/controllers/channel.rs [hunk 2]
```diff
@@ -41,7 +41,7 @@ pub async fn create_channel(
     Json(channel): Json<CreateChannelSchema>,
 ) -> Result<(StatusCode, Json<ChannelModel>), CustomError> {
     if channel.name.is_empty() {
-        return Err(CustomError::BadRequest);
+        return Err(CustomError::BadRequest("Channel name is empty".to_string()));
```
Comment (LukeBrandon) [inline backend/src/controllers/channel.rs]:
```suggestion
        return Err(CustomError::BadRequest("A channel name is required".to_string()));
```

File: backend/src/errors.rs [hunk 3]
```diff
@@ -3,89 +3,94 @@ use axum::{
...
+                } else if pg_error.is_foreign_key_violation() {
+                    return CustomError::BadRequest(format!(
+                        "Foreign violation: {}",
+                        pg_error.message()
+                    ));
+                } else if pg_error.is_unique_violation() {
+                    return CustomError::BadRequest(format!(
+                        "Duplicate key: {}",
+                        pg_error.message()
+                    ));
```
Comment (LukeBrandon) [inline backend/src/errors.rs:50]: else 500?

File: backend/src/errors.rs [hunk 4]
```diff
@@ -3,89 +3,94 @@ use axum::{
...
+                }
+                CustomError::BadRequest("A database error occurred".to_string())
```
Comment (LukeBrandon) [inline backend/src/errors.rs]: Hm? Should the catchall be a 400?

File: backend/src/errors.rs [hunk 5]
```diff
@@ -3,89 +3,94 @@ use axum::{
     response::{IntoResponse, Json, Response},
 };
 use serde_json::json;
-use sqlx::Error as SqlxError;
-use std::fmt;
+use sqlx::{error::DatabaseError, Error as SqlxError};
 
 pub enum CustomError {
-    NotFound(String),
```
Comment (LukeBrandon) [inline backend/src/errors.rs:10]: How are we not using this?

File: backend/src/errors.rs (hunk 4 above)
Comment (FlynnOConnell) [reply backend/src/errors.rs]: BadRequest is a 400, no?

File: backend/src/errors.rs (hunk 5 above)
Comment (FlynnOConnell) [reply backend/src/errors.rs:10]: My thinking was that NotFound is covered in BadRequest

File: backend/src/errors.rs (hunk 4 above)
Comment (FlynnOConnell) [reply backend/src/errors.rs]: Because the From:: block converts `BadRequest` to `http::status::BAD_REQUEST`

Comment (LukeBrandon) [reply backend/src/errors.rs]: I know that.  I mean if we get any other DB error, is it probably a 500? Idk what else would constitute a 400

File: backend/src/errors.rs [hunk 6]
```diff
@@ -3,89 +3,94 @@ use axum::{
...
-                format!("Database Error: {}", db_error),
-            ),
         };
+
         let body = Json(json!({
             "error": error_message,
         }));
 
         (status, body).into_response()
     }
```
Comment (LukeBrandon) [inline backend/src/errors.rs]: where is not found

File: backend/src/errors.rs [hunk 7]
```diff
@@ -3,89 +3,94 @@ use axum::{
...
+/// fn my_function() -> Result<(), CustomError> {
+///     another_function()?;  // This function returns `Result<(), sqlx::Error>`
+///     Ok(())
+/// }
+/// ```
 impl From<SqlxError> for CustomError {
     fn from(err: SqlxError) -> Self {
         match err {
-            SqlxError::Database(e) => CustomError::DatabaseError(sqlx::Error::Database(e)),
+            SqlxError::Database(db_error) => {
```
Comment (LukeBrandon) [inline backend/src/errors.rs:33]: Add RowNoTFound error mapping to a 404

**Review by LukeBrandon — CHANGES_REQUESTED** [review]
add 404.  We should also probably consider changing the name of the error to something better than `CustomerError`.  Maybe `FlukeApiError`?

File: backend/src/errors.rs (hunk 4 above)
Comment (FlynnOConnell) [reply backend/src/errors.rs]: Ah I see, you're saying it should be a 500. That makes sense

File: backend/src/errors.rs (hunk 7 above)
Comment (FlynnOConnell) [reply backend/src/errors.rs:33]: [3d1d7a8](https://github.com/LukeBrandon/fluke/pull/57/commits/3d1d7a8d65afa3e1027e9e7e36ea515a77d99d0f)

File: backend/src/errors.rs (hunk 5 above)
Comment (FlynnOConnell) [reply backend/src/errors.rs:10]: [3d1d7a8](https://github.com/LukeBrandon/fluke/pull/57/commits/3d1d7a8d65afa3e1027e9e7e36ea515a77d99d0f)

**Comment by FlynnOConnell** [comment]
> add 404. We should also probably consider changing the name of the error to something better than `CustomerError`. Maybe `FlukeApiError`?

Added 404 and 401 UNAUTHORIZED for the future user authentication

**Review by LukeBrandon — APPROVED** [review]
im assuming you've tested this lol

**Pull request merged**